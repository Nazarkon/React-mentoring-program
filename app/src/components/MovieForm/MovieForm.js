import React from 'react';
import PropTypes from 'prop-types';

import './MovieForm.scss';

import genreList from '../../mock/genreListData.json';

const MovieForm = ({ movieInfo, handleSubmit }) => {
  const handleSubmitForm = (event) => {
    event.preventDefault();
    const data = Object.fromEntries(new FormData(event.target));
    console.log(data, 'data');
    handleSubmit(data);
  };

  const handleFormReset = (event) => {
    event.preventDefault();
    event.target.form.reset();
    console.log('reset');
    handleSubmit();
  };

  return (
    <form className="movie-form-container" onSubmit={handleSubmitForm}>
      <div className="movie-form-container-input">
        <label className="movie-form-label-container movie-form-label-main">
          <span className="movie-form-label-title">TITLE</span>
          <input
            className="movie-form-label-input"
            id="title"
            name="title"
            type="text"
            placeholder="Provide title"
            defaultValue={movieInfo.title}
          />
        </label>
        <label className="movie-form-label-container-second">
          <span className="movie-form-label-title">Date:</span>
          <input
            className="movie-form-label-input-second"
            id="date"
            type="date"
            name="date"
            placeholder="Select Date"
            defaultValue={movieInfo.releaseDate}
          />
        </label>
      </div>
      <div className="movie-form-container-input">
        <label className="movie-form-label-container movie-form-label-main">
          <span className="movie-form-label-title">MOVIE URL</span>
          <input
            className="movie-form-label-input"
            type="text"
            id="movieURL"
            name="movieURL"
            placeholder="https://"
            defaultValue={movieInfo.movieURL}
          />
        </label>
        <label className="movie-form-label-container">
          <span className="movie-form-label-title">RATING</span>
          <input
            className="movie-form-label-input-second"
            type="text"
            id="movieRating"
            name="movieRating"
            placeholder="0.0"
            defaultValue={movieInfo.movieRating}
          />
        </label>
      </div>
      <div className="movie-form-container-input">
        <label className="movie-form-label-container movie-form-label-main">
          <span className="movie-form-label-title">GENRE</span>
          <select
            className="movie-form-label-input"
            id="genre"
            name="genre"
            defaultValue={movieInfo.genre}
          >
            {genreList.map((genre) => (
              <option key={genre.id} value={genre.id}>
                {genre.name}
              </option>
            ))}
          </select>
        </label>
        <label className="movie-form-label-container">
          <span className="movie-form-label-title">RUNTIME</span>
          <input
            className="movie-form-label-input-second"
            type="text"
            name="runtime"
            id="runtime"
            placeholder="min"
            defaultValue={movieInfo.runtime}
          />
        </label>
      </div>
      <label className="movie-form-label-container">
        <span className="movie-form-label-title">OVERVIEW</span>
        <textarea
          id="overview"
          name="overview"
          className="movie-form-label-textarea"
          defaultValue={movieInfo.overview}
          placeholder="Movie description"
        />
      </label>
      <div className="movie-form-container-buttons">
        <button
          className="movie-form-container-button-cancel"
          type="reset"
          onClick={handleFormReset}
        >
          Cancel
        </button>
        <button className="movie-form-container-button-submit" type="submit">
          Submit
        </button>
      </div>
    </form>
  );
};

MovieForm.propTypes = {
  movieInfo: PropTypes.object.isRequired,
  handleSubmit: PropTypes.func.isRequired
};

export default MovieForm;

MovieForm.defaultProps = {
  movieInfo: {
    title: '',
    releaseDate: '',
    movieURL: '',
    movieRating: '',
    genre: '',
    runtime: '',
    overview: ''
  }
};
