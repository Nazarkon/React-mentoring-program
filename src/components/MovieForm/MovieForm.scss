.movie-form-container {
    max-width: 700px;
    display: flex;
    flex-direction: column;
    .movie-form-container-input {
        display: flex;
        justify-content: space-between;
    }
    .movie-form-label-main{
        margin-right: 30px;
    }
    .movie-form-label-container{
        display: flex;
        flex-direction: column;
        justify-content: flex-start;
        margin-bottom: 10px;
        .movie-form-label-input{
            width: 100%;
            min-width: 500px;
            height: 50px;
            font-size: 20px;
            background: rgba(50, 50, 50, 0.8);
            color: white;
            border: 0;
        }
        .movie-form-label-input-second{
            width: 100%;
            max-width: 140px;
            height: 50px;
            font-size: 20px;
            background: rgba(50, 50, 50, 0.8);
            color: white;
            border: 0;
        }
        .movie-form-label-textarea{
            width: 100%;
            min-width: 500px;
            height: 150px;
            font-size: 20px;
            background: rgba(50, 50, 50, 0.8);
            color: white;
            mix-blend-mode: normal;
            border: 0;
        }
        .movie-form-label-title{
            color: #F65261;
            text-transform: uppercase;
            font-size: 16px;
            margin-bottom: 13px;
        }

        input::placeholder {
            font-size: 20px;
            opacity: 0.5;
            color: white;
        }
    }
    .movie-form-label-container-second{
        display: flex;
        width: 100%;
        max-width: 140px;
        flex-direction: column;
        margin-bottom: 1rem;
    }
    .movie-form-label-title{
        color: #F65261;
        text-transform: uppercase;
        font-size: 16px;
        margin-bottom: 13px;
    }
    .movie-form-container-buttons{
        text-align: right;
        .movie-form-container-button-cancel{
            margin-right: 13px;
            color:#FFFFFF;
            text-transform: uppercase;
            padding: 17px 57px;
            border: 1px solid transparent;
            background-color: #F65261;
            &:hover{
                background-color: transparent;
                color: #F65261;
                border: 1px solid #F65261;
            }
        }
        .movie-form-container-button-submit{
            color:#FFFFFF;
            text-transform: uppercase;
            padding: 17px 57px;
            background-color: #F65261;
            border: 1px solid transparent;
            &:hover{
                background-color: transparent;
                color: #F65261;
                border: 1px solid #F65261;
            }
        }
    }
    input[type="date"]{
        background-color: rgba(50, 50, 50, 0.8);
        padding: 15px;
        font-family: "Roboto Mono",monospace;
        color: #ffffff;
        font-size: 18px;
        border: none;
        outline: none;
        cursor: pointer;
    }
}