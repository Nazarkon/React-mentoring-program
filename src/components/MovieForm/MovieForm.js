import React, { useState, useEffect } from "react";

import "./MovieForm.scss"

import genreList from '../../mock/genreListData.json'

const MovieForm = ({movieInfo , handleSubmit, handleCancel}) => {


        const [title, setTitle] = useState(movieInfo?.title || '');
        const [releaseDate, setReleaseDate] = useState(movieInfo?.releaseDate || '');
        const [movieURL, setMovieURL] = useState(movieInfo?.movieURL || '');
        const [rating, setRating] = useState(movieInfo?.rating || '');
        const [genre, setGenre] = useState(movieInfo?.genre || 'default');
        const [runtime, setRuntime] = useState(movieInfo?.runtime || '');
        const [overview, setOverview] = useState(movieInfo?.overview || '');
        const [formValid, setFormValid] = useState(false);

        useEffect(() => {
            if (
              title !== "" &&
              releaseDate !== "" &&
              movieURL !== "" &&
              rating !== "" &&
              genre !== "default" &&
              runtime !== "" &&
              overview !== ""
            ) {
              setFormValid(true);
            } else {
              setFormValid(false);
            }
          }, [title, releaseDate, movieURL, rating, genre, runtime, overview]);


          const submitForm = (event) => {
            event.preventDefault();
            const formData = {
              title: title,
              releaseDate: releaseDate,
              movieURL: movieURL,
              rating: rating,
              genre: genre,
              runtime: runtime,
              overview: overview
            };
            console.log(formData, 'formData')
            handleSubmit(formData);
          }
        

    return (
        <form className="movie-form-container" onSubmit={(event) => submitForm(event)}>
            <div className="movie-form-container-input">
                <label className="movie-form-label-container movie-form-label-main">
                    <span className="movie-form-label-title">TITLE</span>
                    <input
                    className="movie-form-label-input"
                    type="text"
                    placeholder="Provide title"
                    value={title}
                    onChange={(event) => setTitle(event.target.value)}
                    />
                </label>
                <label className="movie-form-label-container-second">
                <span className="movie-form-label-title">Date:</span>
                    <input
                    className="movie-form-label-input-second"
                    type="date"
                    placeholder="Select Date"
                    value={releaseDate}
                    onChange={(event) => setReleaseDate(event.target.value)}
                    />
                </label>
            </div>
            <div className="movie-form-container-input">
                <label className="movie-form-label-container movie-form-label-main">
                <span className="movie-form-label-title">MOVIE URL</span>
                    <input
                    className="movie-form-label-input"
                    type="text"
                    placeholder="https://"
                    value={movieURL}
                    onChange={(event) => setMovieURL(event.target.value)}
                    />
                </label>
                <label className="movie-form-label-container">
                <span className="movie-form-label-title">RATING</span>
                    <input
                    className="movie-form-label-input-second"
                    type="text"
                    placeholder="0.0"
                    value={rating}
                    onChange={(event) => setRating(event.target.value)}
                    />
                </label>
            </div>
            <div className="movie-form-container-input">
                <label className="movie-form-label-container movie-form-label-main">
                <span className="movie-form-label-title">GENRE</span>
                    <select className="movie-form-label-input" value={genre} onChange={(event) => setGenre(event.target.value)}>
                    {genreList.map((genre) => (
                        <option key={genre.id} value={genre.id}>{genre.name}</option>
                    ))}
                    </select>
                </label>
                <label className="movie-form-label-container">
                <span className="movie-form-label-title">RUNTIME</span>
                    <input
                    className="movie-form-label-input-second"
                    type="text"
                    placeholder="min"
                    value={runtime}
                    onChange={(event) => setRuntime(event.target.value)}
                    />
                </label>
            </div>
                <label className="movie-form-label-container">
                <span className="movie-form-label-title">OVERVIEW</span>
                    <textarea
                    className="movie-form-label-textarea"
                    value={overview}
                    placeholder="Movie description"
                    onChange={(event) => setOverview(event.target.value)}
                    />
                </label>
                <div className="movie-form-container-buttons">
                    <button
                        onClick={handleCancel}
                        className="movie-form-container-button-cancel" 
                        type="button"
                    >Cancel</button>
                    <button
                        disabled={!formValid}
                        className="movie-form-container-button-submit"
                        type="submit"
                    >Submit</button>
                </div>
        </form>
    )
}

export default MovieForm