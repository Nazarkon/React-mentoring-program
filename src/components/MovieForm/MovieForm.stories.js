import React from "react";

import MovieForm from "./MovieForm";

export default {
    title: 'Components/MovieForm',
    component: MovieForm,
    argTypes: {
        movieInfo: {
            content: 'object',
            default: {}
        },
        handleSubmit: {
            action: "clicked"
        }
    },
}

const mockMovieInfo = {
    title: "Film 1",
    releaseDate: '2023-11-12',
    movieURL: 'https://www.overclockers.ua/',
    movieRating: '12.0',
    genre: 'Comedy',
    runtime: '12h',
    overview: 'Lorem Ipsum'
}

const EmptyFormTemplate = (args) => <MovieForm  {...args} />;

const FillWithData = (args) => <MovieForm movieInfo={mockMovieInfo} handleSubmit={args.handleSubmit}/>

export const EmptyForm = EmptyFormTemplate.bind({});
export const FilledForm = FillWithData.bind({});