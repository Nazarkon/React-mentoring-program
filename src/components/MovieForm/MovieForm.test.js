/* eslint-disable testing-library/prefer-screen-queries */
import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import MovieForm from './MovieForm';
import '@testing-library/jest-dom'

describe('MovieForm', () => {
  const handleSubmitMock = jest.fn();
  const handleCancelMock = jest.fn();
  const movieInfo = {
    title: 'The Matrix',
    releaseDate: '1999-03-31',
    movieURL: 'https://movie-test/',
    rating: '8.7',
    genre: 'Comedy',
    runtime: '136',
    overview: 'A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.'
  };

  it("matches snapshot", () => {
    const { container } = render(
      <MovieForm handleSubmit={handleSubmitMock} handleCancel={handleCancelMock} movieInfo={movieInfo} />);
    expect(container).toMatchSnapshot();
  });

  beforeEach(() => {
    handleSubmitMock.mockClear();
    handleCancelMock.mockClear();
  });

  test('Should renders the form fields', () => {
    const {getByLabelText, getByRole}  =  render(<MovieForm handleSubmit={handleSubmitMock} handleCancel={handleCancelMock} movieInfo={movieInfo} />);
    expect(getByLabelText(/title/i)).toBeInTheDocument();
    expect(getByLabelText(/date/i)).toBeInTheDocument();
    expect(getByLabelText(/movie url/i)).toBeInTheDocument();
    expect(getByLabelText(/rating/i)).toBeInTheDocument();
    expect(getByLabelText(/genre/i)).toBeInTheDocument();
    expect(getByLabelText(/runtime/i)).toBeInTheDocument();
    expect(getByLabelText(/overview/i)).toBeInTheDocument();
    expect(getByRole('button', { name: /cancel/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /submit/i })).toBeInTheDocument();
  });

  test('Should submits successfully', async () => {
    const {getByLabelText, getByRole} = render(<MovieForm handleSubmit={handleSubmitMock} handleCancel={handleCancelMock} movieInfo={movieInfo} />);
    userEvent.type(getByLabelText(/title/i), 'Inception');
    userEvent.type(getByLabelText(/date/i), '2010-07-16');
    userEvent.type(getByLabelText(/movie url/i), 'https://movie-test/2');
    userEvent.type(getByLabelText(/rating/i), '8.8');
    userEvent.selectOptions(getByLabelText(/genre/i), 'Comedy');
    userEvent.type(getByLabelText(/runtime/i), '148');
    userEvent.type(getByLabelText(/overview/i), 'A thief who steals corporate secrets through the use of dream-sharing technology is given the inverse task of planting an idea into the mind of a C.E.O.');
    userEvent.click(getByRole('button', { name: /submit/i }));
    expect(handleSubmitMock).toHaveBeenCalledTimes(1);
  });

  test('Should clicks cancel button and calls handleCancel', () => {
    const {getByRole} = render(<MovieForm handleSubmit={handleSubmitMock} handleCancel={handleCancelMock} movieInfo={movieInfo} />);
    userEvent.click(getByRole('button', { name: /cancel/i }));
    expect(handleCancelMock).toHaveBeenCalledTimes(1);
  });
});