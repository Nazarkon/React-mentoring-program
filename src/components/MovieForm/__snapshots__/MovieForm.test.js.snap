// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieForm matches snapshot 1`] = `
<div>
  <form
    class="movie-form-container"
  >
    <div
      class="movie-form-container-input"
    >
      <label
        class="movie-form-label-container movie-form-label-main"
      >
        <span
          class="movie-form-label-title"
        >
          TITLE
        </span>
        <input
          class="movie-form-label-input"
          id="title"
          name="title"
          placeholder="Provide title"
          type="text"
          value="The Matrix"
        />
      </label>
      <label
        class="movie-form-label-container-second"
      >
        <span
          class="movie-form-label-title"
        >
          Date:
        </span>
        <input
          class="movie-form-label-input-second"
          id="date"
          name="date"
          placeholder="Select Date"
          type="date"
          value="1999-03-31"
        />
      </label>
    </div>
    <div
      class="movie-form-container-input"
    >
      <label
        class="movie-form-label-container movie-form-label-main"
      >
        <span
          class="movie-form-label-title"
        >
          MOVIE URL
        </span>
        <input
          class="movie-form-label-input"
          id="movieURL"
          name="movieURL"
          placeholder="https://"
          type="text"
          value="https://movie-test/"
        />
      </label>
      <label
        class="movie-form-label-container"
      >
        <span
          class="movie-form-label-title"
        >
          RATING
        </span>
        <input
          class="movie-form-label-input-second"
          id="movieRating"
          name="movieRating"
          placeholder="0.0"
          type="text"
          value="8.7"
        />
      </label>
    </div>
    <div
      class="movie-form-container-input"
    >
      <label
        class="movie-form-label-container movie-form-label-main"
      >
        <span
          class="movie-form-label-title"
        >
          GENRE
        </span>
        <select
          class="movie-form-label-input"
          id="genre"
          name="genre"
        >
          <option
            value="default"
          >
            Select Genre
          </option>
          <option
            value="documentary"
          >
            Documentary
          </option>
          <option
            selected=""
            value="comedy"
          >
            Comedy
          </option>
          <option
            value="horror"
          >
            Horror
          </option>
          <option
            value="crime"
          >
            Crime
          </option>
        </select>
      </label>
      <label
        class="movie-form-label-container"
      >
        <span
          class="movie-form-label-title"
        >
          RUNTIME
        </span>
        <input
          class="movie-form-label-input-second"
          id="runtime"
          name="runtime"
          placeholder="min"
          type="text"
          value="136"
        />
      </label>
    </div>
    <label
      class="movie-form-label-container"
    >
      <span
        class="movie-form-label-title"
      >
        OVERVIEW
      </span>
      <textarea
        class="movie-form-label-textarea"
        id="overview"
        name="overview"
        placeholder="Movie description"
      >
        A computer hacker learns from mysterious rebels about the true nature of his reality and his role in the war against its controllers.
      </textarea>
    </label>
    <div
      class="movie-form-container-buttons"
    >
      <button
        class="movie-form-container-button-cancel"
        type="reset"
      >
        Cancel
      </button>
      <button
        class="movie-form-container-button-submit"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</div>
`;
